# List of Postal Codes in Toronto Exercise 1 - Segmenting and Clustering Neighbourhoods

## Set up section

#### This first batch of code is to load all the relevent directories etc


In [158]:
import numpy as np # library for vectorized computation
import pandas as pd # library to process data as dataframes

import matplotlib.pyplot as plt # plotting library
# backend for rendering plots within the browser
%matplotlib inline 

from sklearn.cluster import KMeans 
from sklearn.datasets.samples_generator import make_blobs

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors


!conda install -c conda-forge folium=0.5.0 --yes 
import folium 
from urllib import request
import requests
import urllib.request
import time
from bs4 import BeautifulSoup



Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.



In [159]:
import folium # map rendering library
from urllib import request
import requests
import urllib.request
import time
from bs4 import BeautifulSoup
import bs4 as bs


print('Libraries imported.')

Libraries imported.


### This section will scrape the website for the basic details
#### 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.html
#### I used Python BeautifulSoup and Python lxml

In [160]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
rawpage = request.urlopen(url)

In [161]:
def scrape_table_bs4(cname,cols):
    page = urllib.request.urlopen(url).read()
    soup = bs.BeautifulSoup(page,'lxml')
    table = soup.find('table',class_=cname)
    header = [head.findAll(text=True)[0].strip() for head in table.find_all("th")]
    data = [[td.findAll(text=True)[0].strip() for td in tr.find_all("td")] for tr in table.find_all("tr")]
    data = [row for row in data if len(row) == cols]
# This is to store the data temporarily
    temp_df = pd.DataFrame(data, columns=header)
    return temp_df

### This is to parse the data uisng xpath

In [162]:
def scrape_table_lxml(XPATH,cols):
    page = requests.get(url)
    doc = lh.fromstring(page.content)
    table_content = doc.xpath(XPATH)
    for table in table_content: headers = [th.text_content().strip() for th in table.xpath('//th')]
    headers = headers[0:3]
    data = [[td.text_content().strip() for td in tr.xpath('td')] for tr in table.xpath('//tbody/tr')]
    data = [row for row in data if len(row) == cols]
    temp_df = pd.DataFrame(data, columns=headers)
    return temp_df        

### This is to test the work in Beautiful Soup

In [163]:
raw_Postcodes = scrape_table_bs4("wikitable",3)

### This is to test the work in LXML

In [164]:

print ("Postcodes")
print(raw_Postcodes.info(verbose = True))

Postcodes
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287 entries, 0 to 286
Data columns (total 3 columns):
Postcode         287 non-null object
Borough          287 non-null object
Neighbourhood    287 non-null object
dtypes: object(3)
memory usage: 6.8+ KB
None


# Assumptions

#### The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

#### I will only process the cells that have an assigned borough. I will ignore cells with a borough that is Not assigned.

#### Where more than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.

#### If a cell has a borough but a Not assigned neighborhood, then I will make the neighborhood will be the same as the borough. For example the 9th cell in 

In [165]:
Postcodes = raw_Postcodes[~raw_Postcodes['Borough'].isin(['Not assigned'])]
                          
Postcodes=Postcodes.sort_values(by=['Postcode', 'Borough', 'Neighbourhood'], ascending =[1,1,1]).reset_index(drop=True)

In [166]:
Postcodes.loc[Postcodes['Neighbourhood'] == 'Not assigned', ['Neighbourhood']]=Postcodes['Borough']

check_unassigned_post_state_sample = Postcodes.loc[Postcodes['Borough'] =='Queen\'s Park']


In [167]:
Postcodes = Postcodes.groupby(['Postcode','Borough'])['Neighbourhood'].apply(', '.join).reset_index()

### List if Postal Codes in Toronto Canada (starting M...)

In [168]:
Postcodes

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A      Queen's Park   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                       Malvern, Rouge  
1               Highland Creek, Port Union, Rouge Hill  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7          

In [169]:
Postcodes.shape

(103, 3)

# Second Part of Assignment
## Get Coords of each Neighbourhood

In order to utilize the Foursquare location data, I will get the latitude and the longitude coordinates of each neighborhood.
 Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

I chose to use the provided csv file  becuase of   - http://cocl.us/Geospatial_data

In [170]:
lat_longcsv = 'http://cocl.us/Geospatial_data'
!wget -q -o 'Geospatial_coordinates.csv' lat_longcsv
geopostcode_data=pd.read_csv(lat_longcsv).set_index('Postal Code')
geopostcode_data.head()

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476

In [171]:
Postcodes.to_csv('postcode1_df.csv',index=False)

postcode_csv = 'postcode1_df.csv'

postcodes1 = pd.read_csv(postcode_csv).set_index('Postcode')
postcodes1.rename_axis('Postal Code', axis = 'index', inplace = True)
postcodes1.head()


Borough                           Neighbourhood
Postal Code                                                     
M1B          Scarborough                          Malvern, Rouge
M1C          Scarborough  Highland Creek, Port Union, Rouge Hill
M1E          Scarborough       Guildwood, Morningside, West Hill
M1G          Scarborough                                  Woburn
M1H          Scarborough                               Cedarbrae

#### Combine the two dataframes using postal codes as key

In [172]:
Combined_data = postcodes1.join( geopostcode_data)
Combined_data.head()

Borough                           Neighbourhood   Latitude  \
Postal Code                                                                   
M1B          Scarborough                          Malvern, Rouge  43.806686   
M1C          Scarborough  Highland Creek, Port Union, Rouge Hill  43.784535   
M1E          Scarborough       Guildwood, Morningside, West Hill  43.763573   
M1G          Scarborough                                  Woburn  43.770992   
M1H          Scarborough                               Cedarbrae  43.773136   

             Longitude  
Postal Code             
M1B         -79.194353  
M1C         -79.160497  
M1E         -79.188711  
M1G         -79.216917  
M1H         -79.239476

In [173]:
Combined_data.shape

(103, 4)

# Final Part of the Assignment
## Exploring and clustering the neighbourhoods in Toronto

#### To explore the neighbourhoos of selected cities I will use the Foursquare API.  

#### Firstly I will load all the libraries


In [174]:
!conda install -c conda-forge geopy --yes        
!conda install -c conda-forge folium=0.5.0 --yes 

import numpy as np 
import pandas as pd 

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 


from geopy.geocoders import Nominatim 
GeoLocator = Nominatim(user_agent='My-IBMNotebook')

import requests 
from pandas.io.json import json_normalize 


import matplotlib.cm as cm
import matplotlib.colors as colors


from sklearn.cluster import KMeans


import folium 

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


### Use geophy to get the lat/long values of Toronto Canada

In [175]:
address = 'Toronto, Ontario Canada'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto Canada are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.20.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of Toronto Canada are 43.653963, -79.387207.


### Create a map of Toronto with the cities superimposed

In [176]:

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)


for lat, lng, borough, neighborhood in zip(Combined_data['Latitude'], Combined_data['Longitude'], Combined_data['Borough'], Combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#87cefa',
        fill_opacity=0.5,
        parse_html=False).add_to(map_toronto)

In [177]:
map_toronto

###  Assumption

#### For the purpose of te exercise I will  work with only boroughs that contain the word Toronto and then replicate the same analysis that I did with the New York City data.

In [178]:
toronto_data = Combined_data[Combined_data['Borough'].str.contains("Toronto")].reset_index(drop=True)
print(toronto_data.shape)
toronto_data.head()

(38, 4)


Borough                   Neighbourhood   Latitude  Longitude
0     East Toronto                     The Beaches  43.676357 -79.293031
1     East Toronto    Riverdale, The Danforth West  43.679557 -79.352188
2     East Toronto  India Bazaar, The Beaches West  43.668999 -79.315572
3     East Toronto                 Studio District  43.659526 -79.340923
4  Central Toronto                   Lawrence Park  43.728020 -79.388790

#### I will now recreate the map with the markers on it for the neighourhoods

In [179]:

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)


for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Now using the Foursquare API to explore and segment neighborhoods



In [180]:
CLIENT_ID = 'DWE403I3DYSRFXV4VDIAQOSUD1IMFKWNV4LMVNQWSR5CZMDV' # your Foursquare ID
CLIENT_SECRET = 'UND1K2GR13ZF5ZUYY45MAQINZRAGY4IJ2EXBINYW0FAOPGGI' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: DWE403I3DYSRFXV4VDIAQOSUD1IMFKWNV4LMVNQWSR5CZMDV
CLIENT_SECRET:UND1K2GR13ZF5ZUYY45MAQINZRAGY4IJ2EXBINYW0FAOPGGI


### To explore the neighbourhoods in Toronto

#### I will use the same query as for the NY exercise
https://api.foursquare.com/v2/venues/search?
client_id=CLIENT_ID&client_secret=CLIENT_SECRET&ll=LATITUDE,LONGITUDE&v=VERSION&query=QUERY&radius=RADIUS&limit=LIMIT

In [181]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
      
    
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
     
        answer = requests.get(url).json()
        results = answer['response']['groups'][0]['items']
        
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [182]:
Combined_data = toronto_data
toronto_venues = getNearbyVenues(names=Combined_data['Neighbourhood'],
                                   latitudes=Combined_data['Latitude'],
                                   longitudes=Combined_data['Longitude'])

The Beaches
Riverdale, The Danforth West
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
North Midtown, The Annex, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
Bathurst Quay, CN Tower, Harbourfront West, Island airport, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The Junction Sout

In [183]:
Combined_data.shape

(38, 4)

In [184]:
toronto_venues.head()

Neighbourhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  Riverdale, The Danforth West              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

In [185]:
toronto_venues.shape

(836, 7)

#### Noting that 836 venues have been returned, lets check to see how many venues are in each neighbourhood

In [186]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                               
Adelaide, King, Richmond                                               30   
Bathurst Quay, CN Tower, Harbourfront West, Isl...                     16   
Berczy Park                                                            30   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      18   
Cabbagetown, St. James Town                                            30   
Central Bay Street                                                     30   
Chinatown, Grange Park, Kensington Market                              30   
Christie                                                               17   
Church and Wellesley                                                   30   
Commerce Court, Victoria Hotel                                         30   
Davisville                                                             30   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     16   
Design Exchange, Toronto Dominion Centre                               30   
Dovercourt Village, Dufferin                                           15   
First Canadian Place, Underground city                                 30   
Forest Hill North, Forest Hill West                                     5   
Garden District, Ryerson                                               30   
Harbord, University of Toronto                                         30   
Harbourfront                                                           30   
Harbourfront East, Toronto Islands, Union Station                      30   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         21   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               30   
Moore Park, Summerhill East                                             4   
North Midtown, The Annex, Yorkville                                    21   
North Toronto West                                                     21   
Parkdale, Roncesvalles                                                 15   
Riverdale, The Danforth West                                           30   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     30   
St. James Town                                                         30   
Stn A PO Boxes 25 The Esplanade                                        30   
Studio District                                                        30   
The Beaches                                                             4   

                                                    Neighborhood Longitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                                30   
Bathurst Quay, CN Tower, Harbourfront West, Isl...                      16   
Berczy Park                                                             30   
Brockton, Exhibition Place, Parkdale Village                            22   
Business Reply Mail Processing Centre 969 Eastern                       18   
Cabbagetown, St. James Town                                             30   
Central Bay Street                                                      30   
Chinatown, Grange Park, Kensington Market                               30   
Christie                                                                17   
Church and Wellesley                                                    30   
Commerce Court, Victoria Hotel     

#### Checking on unique categories in each area

In [187]:
print('{} unique venue categories have been found.'.format(len(toronto_venues['Venue Category'].unique())))

184 unique venue categories have been found.


## Analysing each neighbourhood
### Using One Hot  encoding
#### One hot encoding is a process by which categorical variables are converted into a form that could be provided to ML algorithms to do a better job in prediction
#### then sort out the presentation of the data

In [188]:
venues_oh = pd.get_dummies(toronto_venues['Venue Category'])


venues_oh['Neighbourhood'] = toronto_venues['Neighbourhood'] 


fixed_columns = [venues_oh.columns[-1]] + list(venues_oh.columns[:-1])
venues_oh =venues_oh[fixed_columns]

venues_oh.head()

Neighbourhood  Airport  Airport Food Court  Airport Gate  \
0                   The Beaches        0                   0             0   
1                   The Beaches        0                   0             0   
2                   The Beaches        0                   0             0   
3                   The Beaches        0                   0             0   
4  Riverdale, The Danforth West        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0         0            0                    0                 0   
1         0            0                    0                 0   
2         0            0                    0                 0   
3         0            0                    0                 0   
4         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Basketball Stadium  Beer Bar  Beer Store  \
0       0     0    0                   0         0           0   
1       0     0    0                   0         0           0   
2       0     0    0                   0         0           0   
3       0     0    0                   0         0           0   
4       0     0    0                   0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Brazilian Restaurant  \
0                   0       0              0          0                     0   
1                   0       0              0          0                     0   
2                   0       0              0          0                     0   
3                   0       0              0          0                     0   
4                   0       0              0          0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Burger Joint  Burrito Place  \
0               0        0                0             0              0   
1               0        0                0             0              0   
2               0        0                0             0              0   
3               0        0                0             0              0   
4               0        0                0             0              0   

   Bus Line  Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0         0        0     0                          0            0   
1         0        0     0                          0            0   
2         0        0     0                          0            0   
3         0        0     0                          0            0   
4         0        0     0                          0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0                   0               0   
4                     0            0                   0               0   

   Church  Cl

# just confirm that no venues have been dropped (check figure is 836)

In [189]:
venues_oh.shape

(836, 185)

#### Lets work out hte aaverage number of the different types of venues for each neighbourhood to see which ones suit me the best

In [190]:
neighbourhoods_grouped = venues_oh.groupby('Neighbourhood').mean().reset_index()

In [191]:
neighbourhoods_grouped

Neighbourhood  Airport  \
0                            Adelaide, King, Richmond   0.0000   
1   Bathurst Quay, CN Tower, Harbourfront West, Is...   0.0625   
2                                         Berczy Park   0.0000   
3        Brockton, Exhibition Place, Parkdale Village   0.0000   
4   Business Reply Mail Processing Centre 969 Eastern   0.0000   
5                         Cabbagetown, St. James Town   0.0000   
6                                  Central Bay Street   0.0000   
7           Chinatown, Grange Park, Kensington Market   0.0000   
8                                            Christie   0.0000   
9                                Church and Wellesley   0.0000   
10                     Commerce Court, Victoria Hotel   0.0000   
11                                         Davisville   0.0000   
12                                   Davisville North   0.0000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...   0.0000   
14           Design Exchange, Toronto Dominion Centre   0.0000   
15                       Dovercourt Village, Dufferin   0.0000   
16             First Canadian Place, Underground city   0.0000   
17                Forest Hill North, Forest Hill West   0.0000   
18                           Garden District, Ryerson   0.0000   
19                     Harbord, University of Toronto   0.0000   
20                                       Harbourfront   0.0000   
21  Harbourfront East, Toronto Islands, Union Station   0.0000   
22                      High Park, The Junction South   0.0000   
23                     India Bazaar, The Beaches West   0.0000   
24                                      Lawrence Park   0.0000   
25                           Little Portugal, Trinity   0.0000   
26                        Moore Park, Summerhill East   0.0000   
27                North Midtown, The Annex, Yorkville   0.0000   
28                                 North Toronto West   0.0000   
29                             Parkdale, Roncesvalles   0.0000   
30                       Riverdale, The Danforth West   0.0000   
31                                           Rosedale   0.0000   
32                                           Roselawn   0.0000   
33                                 Runnymede, Swansea   0.0000   
34                                     St. James Town   0.0000   
35                    Stn A PO Boxes 25 The Esplanade   0.0000   
36                                    Studio District   0.0000   
37                                        The Beaches   0.0000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0               0.0000        0.0000           0.000           0.0000   
1               0.0625        0.0625           0.125           0.1875   
2               0.0000        0.0000           0.000           0.0000   
3               0.0000        0.0000           0.000           0.0000   
4               0.0000        0.0000           0.000           0.0000   
5               0.0000        0.0000           0.000           0.0000   
6               0.0000        0.0000           0.000           0.0000   
7               0.0000        0.0000           0.000           0.0000   
8               0.0000        0.0000           0.000           0.0000   
9               0.0000        0.0000           0.000           0.0000   
10              0.0000        0.0000           0.000           0.0000   
11              0.0000        0.0000           0.000           0.0000   
12              0.0000        0.0000           0.000           0.0000   
13              0.0000        0.0000           0.000           0.0000   
14              0.0000        0.0000           0.000           0.0000   
15              0.0000        0.0000           0.000           0.0000   
16              0.0000        0.0000           0.000           0.0000   
17              0.0000        0.0000           0.000           0.0000   
18              0.0000        0.0000           0.000           0.0000   
19   

#### check for new size in case items get dropped in the future

In [192]:
neighbourhoods_grouped.shape

(38, 185)

#### convert to a panda's dataframe for easier use later on

In [193]:
num_top = 5
for neigh in neighbourhoods_grouped['Neighbourhood']:
    print(""+neigh+"")
    temp = neighbourhoods_grouped[neighbourhoods_grouped['Neighbourhood'] == neigh].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top))
    print('\n')

Adelaide, King, Richmond
              venue  freq
0              Café  0.10
1        Steakhouse  0.10
2             Hotel  0.07
3  Sushi Restaurant  0.07
4  Asian Restaurant  0.07


Bathurst Quay, CN Tower, Harbourfront West, Island airport, King and Spadina, Railway Lands, South Niagara
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3           Airport  0.06
4               Bar  0.06


Berczy Park
                venue  freq
0  Seafood Restaurant  0.07
1        Cocktail Bar  0.07
2         Coffee Shop  0.07
3            Beer Bar  0.07
4      Farmers Market  0.07


Brockton, Exhibition Place, Parkdale Village
            venue  freq
0     Coffee Shop  0.09
1            Café  0.09
2  Breakfast Spot  0.09
3     Yoga Studio  0.05
4          Bakery  0.05


Business Reply Mail Processing Centre 969 Eastern
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2       Garden Center  0.06
3          Comic 

#### intersting but not very easy to understand
#### lets put them in descending order - so I can see the top 20

In [194]:
def return_most_common_venues(row, num_top):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top]

In [195]:
num_top = 20

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


nvs = pd.DataFrame(columns=columns)
nvs['Neighbourhood'] = neighbourhoods_grouped['Neighbourhood']

for ind in np.arange(neighbourhoods_grouped.shape[0]):
    nvs.iloc[ind, 1:] = return_most_common_venues(neighbourhoods_grouped.iloc[ind, :], num_top)

nvs.shape

(38, 21)

In [196]:
nvs.head(20)

Neighbourhood 1st Most Common Venue  \
0                            Adelaide, King, Richmond                  Café   
1   Bathurst Quay, CN Tower, Harbourfront West, Is...       Airport Service   
2                                         Berczy Park    Seafood Restaurant   
3        Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
4   Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
5                         Cabbagetown, St. James Town                  Café   
6                                  Central Bay Street           Coffee Shop   
7           Chinatown, Grange Park, Kensington Market                  Café   
8                                            Christie                  Café   
9                                Church and Wellesley               Gay Bar   
10                     Commerce Court, Victoria Hotel                  Café   
11                                         Davisville          Dessert Shop   
12                                   Davisville North                   Gym   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
14           Design Exchange, Toronto Dominion Centre           Coffee Shop   
15                       Dovercourt Village, Dufferin                Bakery   
16             First Canadian Place, Underground city                  Café   
17                Forest Hill North, Forest Hill West              Bus Line   
18                           Garden District, Ryerson                  Café   
19                     Harbord, University of Toronto                  Café   

    2nd Most Common Venue    3rd Most Common Venue  4th Most Common Venue  \
0              Steakhouse                    Hotel       Asian Restaurant   
1          Airport Lounge         Airport Terminal                Airport   
2          Farmers Market              Coffee Shop           Cocktail Bar   
3          Breakfast Spot                     Café  Performing Arts Venue   
4             Yoga Studio            Auto Workshop                Brewery   
5             Coffee Shop               Restaurant     Italian Restaurant   
6      Italian Restaurant       Chinese Restaurant                   Café   
7   Vietnamese Restaurant  Comfort Food Restaurant     Mexican Restaurant   
8           Grocery Store                     Park                  Diner   
9               Gastropub                    Diner      Indian Restaurant   
10            Coffee Shop               Restaurant                  Hotel   
11            Coffee Shop         Sushi Restaurant            Pizza Place   
12         Clothing Store                     Park         Breakfast Spot   
13                    Pub      Fried Chicken Joint             Restaurant   
14                   Café               Restaurant    Japanese Restaurant   
15               Pharmacy              Supermarket   Gym / Fitness Center   
16            Coffee Shop                Gastropub          Deli / Bodega   
17                   Park            Jewelry Store                  Trail   
18         Clothing Store              Music Venue               Beer Bar   
19                    Bar       Italian Restaurant                 Bakery   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0       Sushi Restaurant   Monument / Landmark                Pizza Place   
1                    Bar           Coffee Shop                      Plane   
2               Beer Bar                Bistro                   Fountain   
3          Burrito Place            Restaurant               Climbing Gym   
4                    Spa                  Park             Farmers Market   
5              Pet Store                   Pub               Liquor Store   
6        Bubble Tea Shop                   Spa                  Gastropub   
7                 Bakery        Farmers Market               Dessert Shop   
8             Baby Store           Candy Store         Athletics & Sp

#### Looks like Adelaide, King, Richmond are good areas for me.
#### Lets see what other neighbourhoods are in that area.

## Clustering the Neighbourhoods
### Using the k means technique to see what other areas are inthe same vicinity
#### Given that there are over 300 neighbourhoods - I'll make 20 clusters

In [197]:
kclusters = 20
neighbourhoodclustering = neighbourhoods_grouped.drop('Neighbourhood',1)
kmeans = KMeans(n_clusters=kclusters, random_state = 1).fit(neighbourhoodclustering)
print(kmeans.labels_[0:12])
print(len(kmeans.labels_))

[14  8  9  0  7  0  0  0 12  0  1  0]
38


#### Confirm data set to be used going forward

In [198]:
Combined_data.head()

Borough                   Neighbourhood   Latitude  Longitude
0     East Toronto                     The Beaches  43.676357 -79.293031
1     East Toronto    Riverdale, The Danforth West  43.679557 -79.352188
2     East Toronto  India Bazaar, The Beaches West  43.668999 -79.315572
3     East Toronto                 Studio District  43.659526 -79.340923
4  Central Toronto                   Lawrence Park  43.728020 -79.388790

In [199]:
Combined_data.shape

(38, 4)

In [200]:
finalised_data = Combined_data

finalised_data['Cluster Labels'] = kmeans.labels_

finalised_data = finalised_data.join(nvs.set_index('Neighbourhood'), on='Neighbourhood')

finalised_data.head()

Borough                   Neighbourhood   Latitude  Longitude  \
0     East Toronto                     The Beaches  43.676357 -79.293031   
1     East Toronto    Riverdale, The Danforth West  43.679557 -79.352188   
2     East Toronto  India Bazaar, The Beaches West  43.668999 -79.315572   
3     East Toronto                 Studio District  43.659526 -79.340923   
4  Central Toronto                   Lawrence Park  43.728020 -79.388790   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0              14     Health Food Store                   Pub   
1               8      Greek Restaurant        Ice Cream Shop   
2               9                  Park        Sandwich Place   
3               0                  Café           Coffee Shop   
4               7              Bus Line                  Park   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail          Neighborhood           Yoga Studio   
1    Italian Restaurant           Yoga Studio        Cosmetics Shop   
2               Brewery                   Pub            Steakhouse   
3   American Restaurant    Italian Restaurant                Bakery   
4           Swim School           Yoga Studio    Falafel Restaurant   

  6th Most Common Venue        7th Most Common Venue  \
0          Dance Studio         Ethiopian Restaurant   
1               Brewery              Bubble Tea Shop   
2    Italian Restaurant             Sushi Restaurant   
3           Yoga Studio               Ice Cream Shop   
4  Ethiopian Restaurant  Eastern European Restaurant   

         8th Most Common Venue 9th Most Common Venue   10th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant                  Dog Run   
1                   Restaurant                   Pub              Coffee Shop   
2            Fish & Chips Shop          Burger Joint           Ice Cream Shop   
3               Sandwich Place             Bookstore  Comfort Food Restaurant   
4          Dumpling Restaurant               Dog Run           Discount Store   

  11th Most Common Venue     12th Most Common Venue 13th Most Common Venue  \
0         Discount Store                      Diner           Dessert Shop   
1                  Diner               Dessert Shop                    Spa   
2      Food & Drink Shop              Movie Theater          Burrito Place   
3            Fish Market  Middle Eastern Restaurant       Stationery Store   
4                  Diner               Dessert Shop          Deli / Bodega   

  14th Most Common Venue   15th Most Common Venue  16th Most Common Venue  \
0          Deli / Bodega                 Creperie        Cuban Restaurant   
1              Juice Bar  Fruit & Vegetable Store  Furniture / Home Store   
2            Pizza Place       Light Rail Station                     Gym   
3                    Bar       Seafood Restaurant             Cheese Shop   
4   Fast Food Restaurant             Dance Studio        Cuban Restaurant   

  17th Most Common Venue 18th Most Common Venue 19th Most Common Venue  \
0         Farmers Market        Coworking Space         Cosmetics Shop   
1          Grocery Store              Bookstore            Pizza Place   
2              Pet Store           Liquor Store   Fast Food Restaurant   
3                Gay Bar        Thai Restaurant        Coworking Space   
4               Creperie        Coworking Space         Cosmetics Shop   

  20th Most Common Venue  
0      Convenience Store  
1                  Trail  
2           Dance Studio  
3           Neighborhood  
4      Convenience Store

#### Lets now see what this looks like on a map so I can see where things are

In [201]:

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)


x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


markers_colors = []
for lat, lon, poi, cluster in zip(finalised_data['Latitude'], finalised_data['Longitude'], finalised_data['Neighbourhood'], finalised_data ['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters